In [ ]:
#import the necessary library

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas.plotting import scatter_matrix
from sklearn.metrics import r2_score
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [ ]:
#load the data set
data=pd.read_csv('50_Startups.csv')
data.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [ ]:
#checking for any null or empty cells in data
data.isna().sum()

R&D Spend          0
Administration     0
Marketing Spend    0
State              0
Profit             0
dtype: int64

In [ ]:
#checking for datatype of columns
data.dtypes

R&D Spend          float64
Administration     float64
Marketing Spend    float64
State               object
Profit             float64
dtype: object

In [ ]:
#assign dependent and independent variables
X = data[["Marketing Spend"]]
y = data["Profit"]

In [ ]:
# Spliting the data set for training and testing in random way
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
#assiging LinearRegression() method to variable for reusability
lr = LinearRegression()

In [ ]:
#pass training dataset into the assigned LinearRegression() method
lr.fit(X_train.values,y_train)

LinearRegression()

In [ ]:
#predict y values based on x column test data
y_pred = lr.predict(X_test)

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


In [ ]:
#print r-sqaure value
round(r2_score( y_test, y_pred),2)

-0.11

In [ ]:
#print intercept score
print(round(lr.intercept_,2))

48923.31


In [ ]:
#checking the prediction
lr.predict(X[0:1])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([182462.19522503])

In [ ]:
# to check datatype of linear regression
type(lr)

sklearn.linear_model._base.LinearRegression

In [ ]:
#serialization
#model persistance (saving and loading trained models)
import pickle

In [ ]:
# save: 'with' is package deals with file handeling, wb- write
with open('model.pkl','wb') as f:
    pickle.dump(lr,f)

In [ ]:
# load: rb-read
with open('model.pkl', 'rb') as f:
    clf2 = pickle.load(f)

In [ ]:
#predict value of y for 1st 5 rows of x
clf2.predict(X[0:5])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:486: UserWarning: X has feature names, but LinearRegression was fitted without feature names
  warnings.warn(


array([182462.19522503, 174569.16142223, 164389.52411328, 157388.28441585,
       152567.58907765])

In [ ]:
# Display the first few predictions alongside the actual values
predictions = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
print(predictions.head())

       Actual      Predicted
13  134307.35  120440.331743
39   81005.76   98457.010516
30   99937.59   74718.086315
45   64926.08   49462.221939
17  125370.37  128906.206651


# **Interpretations Simple Linear Regression Analysis**
Interpretations: Overview of the Simple Linear Regression Model: This model examined the association between the independent variable of marketing spend and the dependent variable of profit. The dataset was divided into training and testing sets, with 20% put aside for testing and the remaining 80% used to train the model.


Parameters for the model:

If marketing spend is zero, the expected profit would be 48923.31 units, according to the equation a) Intercept (𝛽0) = 48923.31. When there is no marketing expenditure, this is the starting profit.

b) According to R-Squared = -0.11, the model is ineffective at explaining the variation in profit. The model's performance is actually poorer than that of a straightforward mean-based prediction for the dependent variable, as indicated by the negative number. According to this, marketing expenditures by themselves are not a significant predictor of profit in this dataset.